### PR TITLE
Corrected encryption + added decryption

### PR DESCRIPTION
**Pull request by francoisWeber**
Corrected encryption process thanks to Paul's code + added a decryption process

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -58,3 +58,4 @@ src/drive/targets/mobilePreview.html
 
 # testcafe data
 testcafe/data/FILES TO TEST
+db/
```
Comment (paultranvan) [inline .gitignore]: It probably be better located below this: https://github.com/Gara64/cozy-drive/blob/fe1751c69d138c55fc10405a22b4bdad49b56dd8/.gitignore#L11

File: src/drive/lib/encryption.js [hunk 2]
```diff
@@ -1,8 +1,16 @@
...
 export const DERIVED_PASSPHRASE_KEY_ID = 'io.cozy.derivedkey.passphrase'
 
+export const encodeArrayBuffer = buff => {
+  return encode(buff)
+}
+
+export const decodeArrayBuffer = str => {
+  return decode(str)
+}
+
```
Comment (paultranvan) [inline src/drive/lib/encryption.js]: These functions could be just replaced with `import { encode as encodeArrayBuffer, decode as decodeArrayBuffer }`

File: src/drive/web/modules/navigation/duck/actions.jsx [hunk 3]
```diff
@@ -25,6 +25,11 @@ import {
 } from 'drive/lib/encryption'
 
 import { ROOT_DIR_ID, TRASH_DIR_ID } from 'drive/constants/config.js'
+import {
+  decodeArrayBuffer,
+  decryptData,
+  importKeyJwk
+} from '../../../../lib/encryption'
```
Comment (paultranvan) [inline src/drive/web/modules/navigation/duck/actions.jsx]: These 3 imports should be grouped with the ones from `'drive/lib/encryption'` 3 lines above ;)

File: src/drive/web/modules/navigation/duck/actions.jsx [hunk 4]
```diff
@@ -431,18 +436,51 @@ export const downloadFiles = files => {
...
 const downloadFile = (file, meta) => {
-  return async dispatch => {
-    const downloadURL = await cozy.client.files
-      .getDownloadLinkById(file.id)
-      .catch(error => {
-        Alerter.error(downloadFileError(error))
-        throw error
-      })
-    const filename = file.name
+  const encrypted = file.metadata.iv !== undefined
```
Comment (paultranvan) [inline src/drive/web/modules/navigation/duck/actions.jsx]: The iv is stored in `file.metadata.encryption` alongside the key: https://github.com/Gara64/cozy-drive/blob/crypto/src/drive/web/modules/upload/index.js#L211

File: src/drive/web/modules/navigation/duck/actions.jsx [hunk 5]
```diff
@@ -431,18 +436,51 @@ export const downloadFiles = files => {
   }
 }
 
+const encryptedDataToBlobURL = async file => {
+  // prepare decryption
+  const encryption = file.metadata
```
Comment (paultranvan) [inline src/drive/web/modules/navigation/duck/actions.jsx]: It should be `file.metadata.encryption`

File: src/drive/web/modules/navigation/duck/actions.jsx [hunk 6]
```diff
@@ -431,18 +436,51 @@ export const downloadFiles = files => {
...
+    }
+  } else {
+    return async dispatch => {
+      const downloadURL = await cozy.client.files
+        .getDownloadLinkById(file.id)
+        .catch(error => {
+          Alerter.error(downloadFileError(error))
+          throw error
+        })
+      // const downloadURL = await decipherData(file)
```
Comment (paultranvan) [inline src/drive/web/modules/navigation/duck/actions.jsx]: Should be removed

File: src/drive/web/modules/navigation/duck/actions.jsx [hunk 7]
```diff
@@ -431,18 +436,51 @@ export const downloadFiles = files => {
...
+      const filename = file.name
 
-    forceFileDownload(`${cozy.client._url}${downloadURL}?Dl=1`, filename)
-    return dispatch({ type: DOWNLOAD_FILE, file, meta })
+      // forceFileDownload(downloadURL, filename)
```
Comment (paultranvan) [inline src/drive/web/modules/navigation/duck/actions.jsx]: Should be removed

File: src/drive/web/modules/upload/index.js [hunk 8]
```diff
@@ -200,17 +196,22 @@ const uploadFile = async (client, file, dirID) => {
     fr.onload = async () => {
       const data = fr.result
       // Encrypt the file with a newly generated AES key
-      const aesKey = await generateAESFileKey()
+      console.log('go aes key')
```
Comment (paultranvan) [inline src/drive/web/modules/upload/index.js]: Should be removed

File: src/drive/web/modules/upload/index.js [hunk 9]
```diff
@@ -200,17 +196,22 @@ const uploadFile = async (client, file, dirID) => {
...
+      const aesKey = await generateAESKey()
+      console.log('go encrypt')
```
Comment (paultranvan) [inline src/drive/web/modules/upload/index.js]: Should be removed

File: src/drive/web/modules/upload/index.js [hunk 10]
```diff
@@ -200,17 +196,22 @@ const uploadFile = async (client, file, dirID) => {
...
       const encrypted = await encryptData(aesKey, data)
+      console.log('go export')
```
Comment (paultranvan) [inline src/drive/web/modules/upload/index.js]: Should be removed

File: src/drive/web/modules/upload/index.js [hunk 11]
```diff
@@ -200,17 +196,22 @@ const uploadFile = async (client, file, dirID) => {
...
       const jwk = await exportKeyJwk(aesKey)
+      console.log('exported : ', jwk)
```
Comment (paultranvan) [inline src/drive/web/modules/upload/index.js]: Should be removed

File: src/drive/web/modules/upload/index.js [hunk 12]
```diff
@@ -200,17 +196,22 @@ const uploadFile = async (client, file, dirID) => {
...
 
       // Create the metadata object containing the encryption info
       // TODO encrypt the key with exportKey
       const encryption = {
         key: jwk,
-        iv: btoa(encrypted.iv)
+        iv: encodeArrayBuffer(encrypted.iv)
       }
-      const name = file.name
+      console.log('encryption : ', encryption)
```
Comment (paultranvan) [inline src/drive/web/modules/upload/index.js]: Should be removed

File: .gitignore (hunk 1 above)
Comment (francoisWeber) [reply .gitignore]: Ups right ! Furthermore, I set up my cozy-stack to use `storage/` directly

File: src/drive/lib/encryption.js (hunk 2 above)
Comment (francoisWeber) [reply src/drive/lib/encryption.js]: I discover this syntax : thanks

File: src/drive/web/modules/navigation/duck/actions.jsx (hunk 4 above)
Comment (francoisWeber) [reply src/drive/web/modules/navigation/duck/actions.jsx]: It was not but now it is in encryption-side and decryption-side

File: src/drive/web/modules/navigation/duck/actions.jsx (hunk 5 above)
Comment (francoisWeber) [reply src/drive/web/modules/navigation/duck/actions.jsx]: cf above: it's corrected everywhere

**Comment by francoisWeber** [comment]
PR updated !

File: src/drive/web/modules/navigation/duck/actions.jsx (hunk 4 above)
Comment (paultranvan) [reply src/drive/web/modules/navigation/duck/actions.jsx]: Shame on me. Thanks!

File: src/drive/web/modules/navigation/duck/actions.jsx (hunk 7 above)
Comment (paultranvan) [reply src/drive/web/modules/navigation/duck/actions.jsx]: Should still be removed ;)

File: src/drive/web/modules/navigation/duck/actions.jsx (hunk 6 above)
Comment (paultranvan) [reply src/drive/web/modules/navigation/duck/actions.jsx]: Should still be removed ;)

**Comment by paultranvan** [comment]
Looks good, could you just remove the 2 comments I pointed?

**Comment by francoisWeber** [comment]
> Looks good, could you just remove the 2 comments I pointed?

Done

**Pull request merged**